src/StructuredLogger/Construction/Construction.cs

```diff
@@ -95,15 +95,23 @@ public void BuildStarted(object sender, BuildStartedEventArgs args)
 
                     EnvironmentFolder = Build.GetOrCreateNodeWithName<Folder>(Intern(Strings.Environment));
 
-                    if (args.BuildEnvironment?.Count > 0)
+                    if (args.BuildEnvironment?.ContainsKey("MSBUILDLOGALLENVIRONMENTVARIABLES") == true)
                     {
                         AddProperties(EnvironmentFolder, args.BuildEnvironment);
                     }
                     else
                     {
+                        if (args.BuildEnvironment?.Count > 0)
+                        {
+                            AddProperties(EnvironmentFolder, args.BuildEnvironment);
+                        }
+
                         EnvironmentFolder.AddChild(new Note
                         {
-                            Text = Intern(Strings.NoEnvironment)
+                            Text = Intern(

```

Comment:
I'm thinking that we don't really need to change anything but the string, if we craft if carefully (and we can reuse the string for NoEnvironment, so no need to introduce a new string for TruncatedEnvironment).

How about:

```
Starting with MSBuild 17.4, only some environment variables are included in the log: the ones read during the build and the ones prefixed with MSBUILD, DOTNET_ or COMPLUS_. Define MSBUILDLOGALLENVIRONMENTVARIABLES to log all environment variables during the build.
```

And leave the previous logic intact (if the dictionary contains items, add them to the tree)

Reply:
and just always add the note to the tree, in all cases

Reply:
Good point; hadn't been thinking both changes would be in 17.4. I'll do that then, thanks!

Reply:
I think it still makes sense to only conditionally add the note—as an indicator to us that the user is logging all environment variables or not if nothing else.

Reply:
I think it’s pretty obvious anyway from the list itself.

We do log the exact MSBuild version that the binlog was created from, so we could do conditional based on that. However that feels flimsy.